### PR TITLE
Composite

### PR DESCRIPTION
**Pull request by edvral**

### CODE REVIEW

File: Server/Models/GameState.cs [hunk 1]
```diff
@@ -177,27 +197,83 @@ public object SendPath()
...
+    {
+        var offsets = new List<(int X, int Y)>();
+        int cols = (int)Math.Ceiling(Math.Sqrt(groupSize));
+        int rows = (int)Math.Ceiling((double)groupSize / cols);
+
+        for (int i = 0; i < groupSize; i++)
+        {
+            int offsetX = (i % cols);
+            int offsetY = (i / cols);
+            offsets.Add((offsetX, offsetY));
```
Comment (vytjar) [inline Server/Models/GameState.cs:273]: Why can't it jsut be `offsets.Add((i % cols, i / cols))`

File: Server/Models/Strategies/ThreatAvoidanceStrategy.cs [hunk 2]
```diff
@@ -108,10 +108,12 @@ private List<PathPoint> GetAdjacentTiles(int x, int y, GameState gameState)
...
             {
-                gameState.Map.GetTile(x + 1, y),
-                gameState.Map.GetTile(x, y + 1),
+                 gameState.Map.GetTile(x + 1, y),
+                 gameState.Map.GetTile(x - 1, y),
+                 gameState.Map.GetTile(x, y + 1),
+                 gameState.Map.GetTile(x, y - 1),
             }
-            .Where(tile => tile != null && tile.Type != TileType.Turret)
+             .Where(tile => tile != null && tile.Type != TileType.Turret)
```
Comment (vytjar) [inline Server/Models/Strategies/ThreatAvoidanceStrategy.cs:116]: Redundant space before period.

File: Server/Models/GameState.cs [hunk 3]
```diff
@@ -298,23 +378,18 @@ public void TowerAttack()
...
-                {
-                    foreach (var shadowEnemy in Map.Enemies.OfType<Enemy>()
-                             .Where(e => e.CurrentStrategy is ShadowStrategy shadowStrategy && shadowStrategy.targetStrongEnemy == strongEnemy))
-                    {
-                        ((ShadowStrategy)shadowEnemy.CurrentStrategy).StopFollowing();
-                    }
-                }
+            if (enemyComponent.IsDead())
+            {
+                Map.Enemies.Remove(enemyComponent);
```
Comment (vytjar) [inline Server/Models/GameState.cs:392]: Enemy is removed from the map at many different places now. What's the point? Removing the enemy right after its death would also reduce the amount of `IsDead()` checks.

File: Server/Models/Strategies/ThreatAvoidanceStrategy.cs [hunk 4]
```diff
@@ -108,10 +108,12 @@ private List<PathPoint> GetAdjacentTiles(int x, int y, GameState gameState)
         {
             return new List<PathPoint>
             {
-                gameState.Map.GetTile(x + 1, y),
-                gameState.Map.GetTile(x, y + 1),
+                 gameState.Map.GetTile(x + 1, y),
```
Comment (vytjar) [inline Server/Models/Strategies/ThreatAvoidanceStrategy.cs:111]: Why was the indentation added?

File: Server/Models/GameState.cs [hunk 5]
```diff
@@ -177,27 +197,83 @@ public object SendPath()
...
+        {
+            int numberOfEnemies = spawnType == 1 ? 2 : 3;
+            var enemyGroup = new EnemyGroup();
+
+            var spawnPositions = GetGroupOffsets(numberOfEnemies);
+
+            foreach (var (offsetX, offsetY) in spawnPositions)
+            {
+                var enemy = CreateAndInitializeEnemy(offsetX, offsetY);
+                enemyGroup.Add(enemy);
```
Comment (vytjar) [inline Server/Models/GameState.cs:218]: `enemyGroup.Add(CreateAndInitializeEnemy(offsetX, offsetY))` ?

File: Server/Models/GameState.cs [hunk 6]
```diff
@@ -111,6 +113,24 @@ public object GetMapEnemies()
             .ToList();
     }
 
+    private IEnumerable<Enemy> GetAllEnemies(IEnumerable<IEnemyComponent> components)
```
Comment (vytjar) [inline Server/Models/GameState.cs:116]: Private method should go below public ones. Additionally the `GameState` class has its methods ordered alphabetically. Would be nice to maintain that.

**Review by vytjar — CHANGES_REQUESTED** [review]

File: Server/Models/GameState.cs (hunk 3 above)
Comment (edvral) [reply Server/Models/GameState.cs:392]: Yes, this is an excessive use of the remove method, but when I remove the remove methods in some places, the groups are not removed from the map or the program does not work as expected.

Comment (vytjar) [reply Server/Models/GameState.cs:299]: Most likely a skill issue. Will approve in that case.

**Review by vytjar — APPROVED** [review]

**Pull request merged**